broadinstitute/single_cell_portal_core
# Migrating Jenkins CD tasks to GitHub Actions (SCP-4944, SCP-5412)

#### BACKGROUND & CHANGES
Starting next May, DevOps will no longer maintain any Jenkins servers in DSP. Since we use Jenkins for our CD pipeline (building Docker images, running deployments), we will need to migrate these tasks over to Github Actions, or we will have to maintain our Jenkins server ourselves.  

This update adds new Github Actions workflows for building Docker images and running deployments to both staging and production.  These new workflows replicate the Jenkins jobs with only minimal changes to the underlying scripts/logic.  Most of the new logic is contained in the workflows and actions themselves, which handle tasks like extracting secrets from vault, and configuring a `gcloud` Docker image that can be used for `ssh` and `scp` access to remote hosts.

All of the new workflows will be able to be triggered manually (note: not available until this PR merges) from the [Actions](https://github.com/broadinstitute/single_cell_portal_core/actions) page.  Additionally, all workflows except the `deploy-to-production` workflow will also trigger automatically using the same conditions as the Jenkins jobs did:
* `build-and-publish-docker-image`: commits to `development` and `master`
* `deploy-to-staging`: completion of `build-and-publish-docker-image` when run on the `development` branch

Additionally, these jobs run on Github-hosted runners, _not_ the self-hosted runner used for CI.  This means these jobs are not affected by backlogs due to multiple PRs.  Access to our staging/production VMs is now federated by Identity Aware Proxy, which allows SCP team members & default service accounts ssh access.  The web interface for staging is still behind the Broad firewall, but portal/database VM access no longer required VPN access.

To do so, add the `--tunnel-through-iap` flag to your ssh command:

```
gcloud compute ssh ubuntu@singlecell-01 --project broad-singlecellportal-staging --zone us-central1-a --tunnel-through-iap
```

NOTE: The last remaining Jenkins job for building the `scp-rails-baseimage` is blocked until SCP-5422 is completed.  This is not a high-priority task, as we only need to run this job once or twice a year.  If we run the job before the Jenkins server is decommissioned, we will have at least 6 months from the decommissioning date to perform the necessary migration.

#### MANUAL TESTING
These cannot be tested manually prior to this PR being merged due to [rules](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) on the `workflow_dispatch` event.  However, you can see evidence of a successful deployment [here](https://github.com/broadinstitute/single_cell_portal_core/actions/runs/7051225191/job/19193635481).  For building the Docker image, there are not any successful jobs left (I accidentally deleted the all) but you can view the remaining [history](https://github.com/broadinstitute/single_cell_portal_core/actions/workflows/build-and-publish-docker-image.yml) to see the successful automation.  The jobs failed because the tag for this branch (`jb-github-actions-cd`) already exists in GCR, which is proof that the existing logic for preventing overwrite of tags other than `development` still works with the new workflows.
